#537-Revamp In-Game-HUD PR

gisketch opened the pull request:
# Revamp In-Game HUD
This PR includes A LOT of new stuff. Mainly visual, and some features. This is my first PR and I started as someone who's not an expert with C# so expect some bad practices I didn't know were actually bad practices. Hopefully, the review will make me learn new stuff and techniques. Well, here goes nothing.

## New Sprites and Fonts
The new sprites of the HUD is a mix between modernly simple and the art style we're going for which is based on the commercial end card. You'll see some of the blades texture will have a little bit of texture in it as well as the gas, etc. Fonts have also been changed to Montserrat which is a beautiful sans-serif font that has a good readability.

## Blades HUD supply
The old blades HUD is working but introduces a spacing bug whenever the resolution changes. That is because the spacing between each blade sprite is controlled in the code so the spacing in a 1080p resolution is the same with that of a 480p resolution but is very noticeable since the value is the same. (ex. 16px of 1080p is very small in comparison with 16px of 480p)
(Issue #522 introduced this bug, so this PR closes #522)

To fix this problem, I used a Unity Component called horizontal lay-out group for each of the remaining blades so that the spacing will always be consistent and is relative to the resolution. I've also made some minor changes to the code to fit in with the layout the art team has worked on.

![image](https://user-images.githubusercontent.com/78424395/112599642-305c9f80-8e4b-11eb-84c8-2d5528691b5c.png)

## Visual Feedback and Notification
Whenever you kill a titan with over a thousand damage, a blood overlay will be instantiated to your screen. I have made 5 blood sprites so far that will be randomly picked whenever the overlay is instantiated.
![Unity_8DaPn1ygO2](https://user-images.githubusercontent.com/78424395/112600794-95fd5b80-8e4c-11eb-8ce2-a166aca1daea.jpg)

There is also a subtle blood overlay on the blade HUD whenever you hit a flesh. The blades will also break into segments depending on the blade's stamina. There is also a low gas warning that shows whenever the gas is below 20%.
![image](https://user-images.githubusercontent.com/78424395/112601108-02785a80-8e4d-11eb-97a9-44d1586ca964.png)

## HUD Customization (Drag & Drop)
This one is a huge one because this is the one that I, with the help of @nolanrdavenport, made the code from scratch and is a huge feature. This introduces the customization of the HUD elements.

The current elements that you can move around and scale are:

- Chatroom
- Top Text
- Topleft Text
- Topright Text
- Compass
- Supply
- Killfeed
- Minimap

This feature is controlled by two scripts. `ChangeHUDHandler.cs` and `CustomizableHUDElement.cs`.

### ChangeHUDHandler.cs
The `ChangeHUDHandler.cs` has different methods that controls the layout such as `SaveHUDLayout(), LoadCustomHUDLayout(), CancelChanges()`. It is also the core script of the feature.

### CustomizableHUDElement.cs
The `CustomizableHUDElement.cs` is attached to each component. Each component has an animator controller to change its appearance whenever there is a need whenever it enters the edit mode. Examples would be the chatroom (changes to show a dummy text) and the killfeed (to show various damages in the killfeed) as they are not by default, visible (See screenshot below). The way it is handled I believe is inefficient and I think there is a better way to handle that. I just do not know how. 

The script also contains 4 pointer event data (e.g., OnPointerEnter, OnPointerExit, OnPointerUp, and OnPointerDown). Before, the HUD elements are moved by clicking a certain radius of an element from its center. I removed that and changed it by using pointer event data so that it uses a raycast for each elements meaning that you don't have to manually get to the center of the element but instead, it will work as long as you click inside the element's bounding box.

The script is also responsible for each of its positions, scales, and visibility's PlayerPrefs. Resolution scaling is also considered in the script so changing resolutions should hopefully still work well.

![image](https://user-images.githubusercontent.com/78424395/112603743-5fc1db00-8e50-11eb-8ae9-126c0db1c473.png)

## Compass

This feature is also controlled by two scripts. `CompassController.cs` and `CompassMarker.cs`

### CompassController.cs
This script is responsible for handling the position of the compass depending on the player's position. As well as the methods for creating markers into a list. The image of the compass is also a raw image to enable repeating texture. It also has a float of max distance for non-global markers.

The only issue I have with the way it's working right now is how it references the MainCamera. Because we don't have an OnSpawn callback, instead of finding the camera on this script. I put it on the IN_GAME_MAIN_CAMERA.cs Start() method. 

![image](https://user-images.githubusercontent.com/78424395/112607304-078cd800-8e54-11eb-8293-c9e867f1aff5.png)


![image](https://user-images.githubusercontent.com/78424395/112605099-e4612900-8e51-11eb-838e-f4640cf8ad1f.png)

### CompassMarker.cs
This script is attached to the GameObject that needs to show a marker in the compass. As you can see there are multiple bools in its settings.

- isFlare (specifically used for flare as #33 changed how the flare works and uses one flare prefab instead of 3 different flares. this setting will allow the script to change its color depending on its attached particle system's color //this definitely could be changed how it's implemented)
- isGlobal (ignores CompassController's maxDistance so it shows anywhere in the map)
- followAlways (follows the position of the gameobject forever instead of only registering the start position, true on titans, false on flares)

![image](https://user-images.githubusercontent.com/78424395/112605376-35711d00-8e52-11eb-9ba3-68b079183d0b.png)

## Hide HUD toggle (by @FAStre)
You can now hide the whole HUD by pressing F1. 

### Minor changes
- Speedometer added
- Scoreboard implementation started (disabled for now, preferably to be continued on a different branch)
- Killfeed animations

### Bugfixes
- Remaining blades do not scale properly #522
- Kill feed scaling issue #167 (fixed by @nolanrdavenport so this PR closes #167)

### End
That's all for my branch. Huge thanks to the art team for providing some of the assets and layouts. Also to @nolanrdavenport for helping me start the branch. This is my first one so it probably consists of bad practices but hopefully the review will help me get better!

ghegi committed:
Converted all UI texts to TextMeshPro

FAStre committed:
Added HideHUD toggle with default bind "F1"

-Allso changed screenshot to be default F2, seems more user friendly(art team also recommended)

ghegi committed:
fixed bug with the ui scaling of blades

nolanrdavenport committed:
Fixed kill feed scaling issue.

The position of the kill feed should now be in the correct position, no matter what the resolution is. Also, the variable "startPositionRatio" holds the proportion of the screen that is below the kill feed. This makes it easy to adjust the location in the future.

ghegi committed:
New gas canister, blade, and chatcontent sprites.

nolanrdavenport committed:
Made the start to having a leaderboard menu

ghegi committed:
leaderboard implementation

ghegi committed:
Leaderboard implemented with default hotkey Tab,

Interaction Wheel is now I

ghegi committed:
Sorting players in the leaderboard now works - sorted by kills by default

ghegi committed:
Damage to TMPro and Crosshair Sprite changes

ghegi committed:
revamped chatbox UI

ghegi committed:
Merge branch 'development' into #537-revamp-ingame-hud

ghegi committed:
Merge branch 'development' into #537-revamp-ingame-hud

ghegi committed:
bloodsmear implementation

nolanrdavenport committed:
Started working on HUD customization

nolanrdavenport committed:
fixed order

ghegi committed:
drag and drop custom hud layout with save and load buttons implemented

nolanrdavenport committed:
Fixed custom HUD loading.

The custom HUD didn't load by default and you had to manually load it from the Change HUD Screen. Now, if you created a custom HUD, it loads by default when you load up the game.

I also prettied up the code

ghegi committed:
Rescaling of HUD elements implemented

ghegi committed:
Toggling of element's visibility now works

this should wrap up the HUD customization

ghegi committed:
Hide hook indicators implemented (default keybind H)

ghegi committed:
Blades HUD revamped with new UI animation for reloading

ghegi committed:
blades UI refined

ghegi committed:
scoreboard scoring implemented (svork)

ghegi committed:
scoreboard is now scrollable

ghegi committed:
sorting is not togglable through (score,kills,dmg)

nolanrdavenport committed:
Improved performance of leaderboard

The leaderboard was being updated every time Update was called. This dropped the frames by over 100. I made it to where it was updated only 8 times a second.

nolanrdavenport committed:
Added ping to leader board.

nolanrdavenport committed:
Added some space between playerId and their message.

nolanrdavenport committed:
Performance work

ghegi committed:
scaling hud bug fix

ghegi committed:
bug fixes for hud saving and loading and new anim controllers for hud elements

ghegi committed:
HUD customization beautified

ghegi committed:
visual improvements

ghegi committed:
sorting order fix

ghegi committed:
Added compass (incomplete)

ghegi committed:
compass fix

ghegi committed:
added compass as a customizable HUD element

ghegi committed:
scoreboard sorting indicator added

nolanrdavenport committed:
Debug work

nolanrdavenport committed:
Update AoTTG 2.unity

nolanrdavenport committed:
Fixed a few bugs

Issue with labels when there is resolution change.

Issue with not being able to spawn if HUD is disabled.

Issue with dragging HUD elements off screen.

ghegi committed:
chat overlapping bug  on change HUD fixed and renamed leaderboard.cs to scoreboard.cs

ghegi committed:
Merge branch '#537-revamp-ingame-hud' of https://github.com/AoTTG-2/AoTTG-2 into #537-revamp-ingame-hud

nolanrdavenport committed:
Fixed killfeed not showing up in singleplayer issue

nolanrdavenport committed:
Merge branch '#537-revamp-ingame-hud' of https://github.com/AoTTG-2/AoTTG-2 into #537-revamp-ingame-hud

nolanrdavenport committed:
Fixed the Canvas Prefab to make it work with scaling better.

nolanrdavenport committed:
Fixed an issue that I caused :D

nolanrdavenport committed:
Weird numbers fixed

ghegi committed:
killfeed is now customizable

ghegi committed:
converted all event triggers to code

ghegi committed:
Compass markers added for flares

ghegi committed:
fixed bug when compass markers stay forever

ghegi committed:
added unique id for each markers

 to avoid duplicate markers with the same name to be destroyed at the same time

ghegi committed:
moving camera on menu bugfix

ghegi committed:
fixed issue from #367

ghegi committed:
new flare marker designs

ghegi committed:
revamped compass container look

ghegi committed:
compass additional fixes

supply now has a marker
titans can now be shown in the compass when nearby

ghegi committed:
fixed bug with being able to open the pause menu while on change hud screen

ghegi committed:
reworked scoreboard

ghegi committed:
make messages separate objects.

ghegi committed:
inroomchat bugfix

ghegi committed:
reworked HUD element's visibility

sets the element's scale to 0 instead of disabling them to avoid nullreference exceptions from elements when loaded.

ghegi committed:
Merge branch 'development' into #537-revamp-ingame-hud

ghegi committed:
minimap confict fixes

ghegi committed:
Merge branch 'development' into #537-revamp-ingame-hud

ghegi committed:
Original Camera Mode pause bugfix.

camera rotates on the x axis when on pause. this commit fixes that bug.

Commit:
UX enhancement for HUD customization

Selected element's selection box will now persist when selected.

Commit:
Blade HUD bugfixes

Commit:
minor visual fixes

Commit:
Blade HUD shake animations + visual enhancements

Commit:
KillFeed animation + titan compass marker enhancements

Commit:
Blood overlay changes + gas notification

Blood smears instantiate random blood overlays instead of one single bloodsmear.
Added gas notification when gas is below 10 percent.

Commit:
Changed gas notification threshold to 20%

Commit:
hook bugfix + hook indicators playerprefs

when hiding hook indicators, it is now setting the gameobjects's scale to Vector3.zero instead of disabling to avoid null references. Their scale is now also saved in PlayerPrefs.

Commit:
killfeed hitbox adjustments for hud customization

Commit:
blood overlays now destroy on disconnect.

Commit:
typo fix (lay-out to layout)

Commit:
messages no longer persist across lobbies

Commit:
reworked RemoveMessageIfMoreThanMax() from InRoomChat.cs

Commit:
scoreboard - autosort score by default

Commit:
code enhancement + comments removal

Commit:
made HUD into its own prefab

Commit:
Merge branch 'development' into #537-revamp-ingame-hud

Commit:
editor readability

Commit:
m

Commit:
Merge branch 'development' into #537-revamp-ingame-hud

Commit:
added made prefabs before conflicts

Commit:
Conflict fixes

Commit:
Customize HUD bugfix

after merging with 33, customize HUD is not useable. this commit fixes it.

Commit:
Fixed crosshair bug

Commit:
temp bladeSta and gas references for supply HUD

Commit:
InRoomChat fix

Commit:
Animation Changes + Compass Marker

- new killfeed animation
- compass marker now has an isFlare bool that when true, checks the particle system's color to define its own color
- removed toggling of doublehook for now

Commit:
Customizable HUD Elements not checking on start bugfix

Commit:
Temporarily disabled scoreboard + readded gas notifications

Commit:
compass markers revamped (again)

Commit:
much more evil titan marker

Commit:
Speedometer implementation

Commit:
Merge branch 'development' into #537-revamp-ingame-hud

Commit:
removed UI blur (shader incompatibility)

Commit:
speedometer hides when paused

FAStre committed:
Re-add Correct HUD toggle

Commit:
Code smell fixes

Commit:
bugfixes introduced by last commit + reversed gas visual

Commit:
Bugfixes

Commit:
code smells fix

Lithrun commented on Assets/Scripts/Characters/Humans/Hero.cs:
```diff
@@ -2058,7 +2058,7 @@ public void Initialize(CharacterPreset preset)
 
             EquipmentType = preset.CurrentBuild.Equipment;
-            Equipment.Initialize();
+                                            Equipment.Initialize();
```
Incorrect formatting, use Ctrl K + D (Ctrl E + D) to automatically fix this

Lithrun commented on Assets/Scripts/Characters/Humans/Hero.cs:
```diff
@@ -3926,6 +3927,7 @@ private void ShowAimUI()
                     distance += "\n" + ((currentSpeed / 100f)).ToString("F1") + "K";
                 }
                 hookUI.distanceLabel.text = distance;
+                hookUI.speedLabel.text = ((currentSpeed / 100f)).ToString("F1") + "K";
```
Don't forget to use the Units / second speedlabel

Lithrun commented on Assets/Scripts/Characters/Humans/Hero.cs:
```diff
@@ -3967,21 +3969,8 @@ private void ShowAimUI()
-            {
-                cachedSprites["GasLeft"].color = cachedSprites["GasRight"].color = Color.white;
-            }
+            cachedSprites["GasLeft"].fillAmount = cachedSprites["GasRight"].fillAmount = 1 - (currentGas / totalGas);
```
Are we still using this cachedSprites? If so, move this magic strings into constants.

Lithrun commented on Assets/Scripts/Characters/PhotonPlayerProperty.cs:
```diff
@@ -49,5 +49,6 @@ public class PhotonPlayerProperty
     public static string team = "team";
     public static string total_dmg = "total_dmg";
     public static string uniform_type = "uniform_type";
+    public static string ping = "ping";
```
How often will this ping player property be updated?

Lithrun commented on Assets/Scripts/DestroyOnDisconnect.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Assets.Scripts.Utility
+{
+    public class DestroyOnDisconnect : PunBehaviour
```
By default, rooms have "PhotonNetwork.autoCleanUpPlayerObjects" enabled. This will automatically do above https://doc-api.photonengine.com/en/pun/v1/class_photon_network.html#a6c6fb3cd57d7e2a13d1fc354db0c1fd7

Lithrun commented on Assets/Scripts/FengGameManagerMKII.cs:
```diff
@@ -353,6 +355,19 @@ private IEnumerator clearlevelE(string[] skybox)
         private void LateUpdate()
         {
+            // Update the players ping once every second. 
+            float timeSince = Time.time * 1000;
```
Such logic does not belong in the FengGameManager. Rather extend the PhotonService for such information

Lithrun commented on Assets/Scripts/Legacy/RCextensions.cs:
```diff
@@ -192,5 +192,22 @@ public static string returnStringFromObject(object obj)
     }
+
+    //Converter for PlayerPrefs
+    public static int boolToInt(bool val)
```
RCExtensions is obsolete due to poor code practices. Any of its logic should be migrated to a new extensions service. Convert.ToInt32(bool) already exists in C#.

Lithrun commented on Assets/Scripts/Legacy/RCextensions.cs:
```diff
@@ -192,5 +192,22 @@ public static string returnStringFromObject(object obj)
+            return 0;
+    }
+
+    public static bool intToBool(int val)
```
Convert.ToBoolean(int) is a default C# method, so this can be removed as well.

Lithrun commented on Assets/Scripts/Room/Chat/InRoomChat.cs:
```diff
@@ -18,9 +18,13 @@ public class InRoomChat : Photon.MonoBehaviour, IUiElement
-        public Text ChatText;
+        public TMP_InputField ChatInputField;
+        public TMP_Text ChatText;
+        public GameObject messagePrefab;
```
public properties and fields are always **P**ascal**C**ase

Lithrun commented on Assets/Scripts/Room/Chat/InRoomChat.cs:
```diff
@@ -18,9 +18,13 @@ public class InRoomChat : Photon.MonoBehaviour, IUiElement
+        public GameObject messagePrefab;
+        public GameObject messagePrefabParent;
         private bool IsChatOpen { get; set; }
+        Regex openingTag = new Regex(@"<([a-z]*)(?:=.+?)?>");
```
Access modifier is missing

Lithrun commented on Assets/Scripts/Room/Chat/InRoomChat.cs:
```diff
@@ -18,9 +18,13 @@ public class InRoomChat : Photon.MonoBehaviour, IUiElement
+        public GameObject messagePrefabParent;
         private bool IsChatOpen { get; set; }
+        Regex openingTag = new Regex(@"<([a-z]*)(?:=.+?)?>");
+        Regex closingTag = new Regex( @"</([a-z]*)>");
```
Access modifier is missing

Lithrun commented on Assets/Scripts/Room/Chat/InRoomChat.cs:
```diff
@@ -144,7 +154,7 @@ private void HandleChatInput(InRoomChat chat)
                         else
                         {
-                            ChatAll(chat.inputLine);
+                            ChatAll("  "+chat.inputLine); // Two spaces to separate playerId and their message. 
```
Rather make these double whitespace a constant, or use string interpolation `$"  {chat.inputLine)";`

Lithrun commented on Assets/Scripts/Room/Chat/InRoomChat.cs:
```diff
@@ -185,8 +185,8 @@ private void CommandHandler(string input)
     {
-        var openingTags = Regex.Matches(message, @"<([a-z]*)(?:=.+?)?>");
-        var closingTags =  Regex.Matches(message, @"</([a-z]*)>");
+        var openingTags = this.openingTag.Matches(message);
```
Avoid using `this.` keyword unless it's explicitly required

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -6,10 +6,12 @@
 using Random = UnityEngine.Random;
+using Assets.Scripts.UI.InGame.HUD;
 
 public class IN_GAME_MAIN_CAMERA : MonoBehaviour
```
Formatting of this file appears to be partially incorrect now.

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -670,8 +671,18 @@ public static void ToggleSpecMode()
+        if (HUD.GetComponent<HUD>().isActive)
+        {
+            HUD.GetComponent<HUD>().isActive = false;
+            HUD.transform.localScale = new Vector3(0, 0, 0);
```
Rather disable the HUD instead of making its scale zero. The HUD will still be rendered and its logic is still running, which wastes performance. Additionally, if any script were to use the HUDs localscale, it will receive a "DivideByZeroException"

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -846,18 +880,24 @@ private void ShakeUpdate()
         GameObject.Find("MultiplayerManager").GetComponent<FengGameManagerMKII>().addCamera(this);
         isPausing = false;
-
+        HUD =GameObject.Find("HUD");
```
Avoid using GameObject.Find, as via the UIService, you should have a reference to the HUD already

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -846,18 +880,24 @@ private void ShakeUpdate()
         CreateSnapShotRT2();
+
+        // Find the compass gameobject to set the camera's transform to this.
+        CompassController compass = GameObject.Find("Compass").GetComponent<CompassController>();
```
Same as HUD, make use of references instead of attempting to find something

Lithrun commented on Assets/Scripts/UI/Camera/IN_GAME_MAIN_CAMERA.cs:
```diff
@@ -846,18 +880,24 @@ private void ShakeUpdate()
     private void OnDestroy()
     {
         EntityService.OnRegister -= EntityService_OnRegistered;
+        GameObject.Find("Compass").GetComponent<CompassController>().compassMode = false;
```
And here as well

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+    {
+        public InGameMenu menu;
+        public HUD.HUD HUD;
+        public GameObject[] HUDelements;
```
HUDElements

Lithrun replied:
Should also be of type: CustomizableHudElement[] then

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+
+namespace Assets.Scripts.UI.InGame
+{
+    public class ChangeHudHandler : UiMenu
```
Again, public properties or fields are always PascalCase

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+        public Toggle toggleVisibility;
+        public bool hasChanged = false;
+
+        protected override void OnEnable()
```
Useless override

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+            base.OnEnable();
+        }
+
+        protected override void OnDisable() {
```
useless override

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+
+        public void Update()
+        {
+            if(HUD.inEditMode)
```
Invert if statements to make the code more readable

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+        public HUD.HUD HUD;
+        public GameObject[] HUDelements;
+        public bool elementSelected = false;
+        public GameObject selectedElement;
```
Rather use CustomizableHudElement instead of GameObject, so you don't have to use GetComponent<CustomizableHudElement> a lot of times

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+        {
+            HUD.inEditMode = true;
+
+            foreach(GameObject element in HUDelements)
```
Now that HUDElements if of type CustomizableHudElement, you don't have to call all those ugly GetComponent<> anymore. If you still had to, then it should only be used once, and then stored as a variable.

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+                element.GetComponent<CustomizableHudElement>().SaveTempPosition();
+            }
+            
+            this.Show();
```
Remove redundant `this.` usage

Lithrun commented on Assets/Scripts/UI/InGame/HUD/ChangeHUDHandler.cs:
```diff
@@ -0,0 +1,143 @@
+                    element.GetComponent<CustomizableHudElement>().StopCustomization();
+                }
+
+                PlayerPrefs.SetInt("hasCustomHUD", 1);
```
Magic string. Make this a constant

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CompassController.cs:
```diff
@@ -0,0 +1,94 @@
+
+namespace Assets.Scripts.UI.InGame.HUD
+{
+    public class CompassController : UiElement
```
PascalCase

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CompassController.cs:
```diff
@@ -0,0 +1,94 @@
+
+
+
+        float compassUnit;
```
Missing access modifier

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CompassController.cs:
```diff
@@ -0,0 +1,94 @@
+
+        float compassUnit;
+
+        void Start()
```
Missing access modifier

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CompassController.cs:
```diff
@@ -0,0 +1,94 @@
+            maxDistance = 150f; //minimap max distance
+        }
+
+        void Update()
```
Missing access modifier

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CompassController.cs:
```diff
@@ -0,0 +1,94 @@
+
+        void Update()
+        {
+            if(compassMode)
```
Invert if statement to make this code more readable.

`if(!CompassMode) return;`
Now the remaining 20 lines of code, don't have to be within that if statement

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CompassController.cs:
```diff
@@ -0,0 +1,94 @@
+            compassMarkers.Remove(marker);
+        }
+
+        Vector2 GetPosOnCompass (CompassMarker marker)
```
Missing access modifier

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CompassController.cs:
```diff
@@ -0,0 +1,94 @@
+
+        Vector2 GetPosOnCompass (CompassMarker marker)
+        {
+            Vector2 playerPos = new Vector2(cam.transform.position.x, cam.transform.position.z);
```
Use var of the right hand side is obvious

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CompassMarker.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace Assets.Scripts.UI.InGame.HUD
+{
+    public class CompassMarker : MonoBehaviour
```
PascalCasing

Lithrun replied:
Missing access modifiers for all methods

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CustomizableHUDElement.cs:
```diff
@@ -0,0 +1,243 @@
+using UnityEngine;
+using Assets.Scripts.UI.InGame;
+
+/* Any HUD element that is customizable should have this script attached */
```
Use `///sumary` above the class with that information, as this comment is fairly easy to miss and thus useless

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CustomizableHUDElement.cs:
```diff
@@ -0,0 +1,243 @@
+namespace Assets.Scripts.UI.InGame.HUD
+{
+
+    public class CustomizableHudElement : MonoBehaviour, IPointerEnterHandler, IPointerExitHandler, IPointerUpHandler, IPointerDownHandler
```
PascalCasing & access modifiers are missing

Lithrun commented on Assets/Scripts/UI/InGame/HUD/CustomizableHUDElement.cs:
```diff
@@ -0,0 +1,243 @@
+
+        public void SavePosition()
+        {
+            PlayerPrefs.SetFloat(PlayerPrefsKey + "CustomX", this.transform.position.x);
```
A lot of magic strings

Lithrun commented on Assets/Scripts/UI/InGame/HUD/DestroySelf.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Assets.Scripts.Utility
+{
+    public class DestroySelf : MonoBehaviour
```
Rather use "SelfDestroy", which already exists and does exactly the same

Lithrun commented on Assets/Scripts/UI/InGame/HUD/GasNotification.cs:
```diff
@@ -0,0 +1,18 @@
+        public Assets.Scripts.UI.InGame.Weapon.Blades blades;
+        public GameObject[] gasNotifications;
+
+        void Update()
```
Access modifier

Lithrun commented on Assets/Scripts/UI/InGame/HUD/HUD.cs:
```diff
@@ -1,41 +1,67 @@
+        
+        public void Update()
+        {
+            if(UnityEngine.Input.GetKeyDown(KeyCode.J))
```
Don't forget to remove this when you are done

Lithrun commented on Assets/Scripts/UI/InGame/HUD/HUD.cs:
```diff
@@ -1,41 +1,67 @@
             foreach (var label in damageLabels)
             {
-                label.text = string.Empty;
+                label.text = "";
```
string.Empty == "", so rather use string.Empty

Lithrun commented on Assets/Scripts/UI/InGame/HUD/HUD.cs:
```diff
@@ -1,41 +1,67 @@
 
+        private void ShowBloodSmear(int damage)
+        {
+            if(damage >= 1000)
```
Why would a BloodSmear effect only appear above 1k damage? With +1k damage you already have the "reward" of a different color damage number. Rather make this 500

Lithrun commented on Assets/Scripts/UI/InGame/HUD/KillInfo.cs:
```diff
@@ -1,26 +1,33 @@
         private float timeElapsed;
 
+        [Header("Colors")] //Customize in the inspector
+        public Color fiveHundred;
```
Rather use ScriptableObjects instead of hard coding this. Name it "DamageColor" or something similar. Within this ScriptableObject, including a Color, and a minimum damage number.

KillInfo will then contain a list of "DamageColors". The order of these DamageColors should be from low to high, then based on a quick loop, you can get the right DamageColor scriptableObject, and thus its color.

Lithrun commented on Assets/Scripts/UI/InGame/HUD/KillInfo.cs:
```diff
@@ -30,73 +37,42 @@ public void MoveOn()
-                {
-                    DamageLabel.color = Color.red;
-                }
+                if (dmg >= 500 && dmg < 1000)
```
This could be done in two lines of code, once moved to scriptableObjects

Lithrun commented on Assets/Scripts/UI/InGame/HUD/Weapon/Blades.cs:
```diff
@@ -1,17 +1,66 @@
 
 namespace Assets.Scripts.UI.InGame.Weapon
 {
     public class Blades : MonoBehaviour
```
PascalCasing & Access Modifiers are your best friends~

Lithrun commented on Assets/Scripts/UI/InGame/HUD/Weapon/Blades.cs:
```diff
@@ -1,17 +1,66 @@
 
+        private void Update()
+        {   
+            if(leftImage != null && rightImage != null)
```
Invert if statement to make it easier to understand. Might want to use a ScriptableObject array as well here, similar to the aforementioned DamageColors

Lithrun commented on Assets/Scripts/UI/InGame/HUD/Weapon/Blades.cs:
```diff
@@ -1,17 +1,66 @@
+        
+        public void ShakeBlades()
+        {
+            if(curLeft.transform.childCount < 5) Instantiate(bloodPrefab, curLeft.transform); //To prevent overpopulating blood prefabs
```
Avoid hard coding this to 5, as this number could change.

Lithrun commented on Assets/Scripts/UI/InGame/HUD/Weapon/Blades.cs:
```diff
@@ -1,17 +1,66 @@
+        {
+            if(curLeft.transform.childCount < 5) Instantiate(bloodPrefab, curLeft.transform); //To prevent overpopulating blood prefabs
+            if(curRight.transform.childCount < 5) Instantiate(bloodPrefab, curRight.transform);
+            currentLeftBlade.GetComponent<Animator>().SetTrigger("Shake");
```
Magic string

Lithrun commented on Assets/Scripts/UI/InGame/Scoreboard/Scoreboard.cs:
```diff
@@ -0,0 +1,195 @@
+        private float timeLast = 0;
+
+
+        protected override void OnEnable()
```
Useless overrides

Lithrun commented on Assets/Scripts/UI/InGame/Scoreboard/Scoreboard.cs:
```diff
@@ -0,0 +1,195 @@
+
+        private void LateUpdate()
+        {
+            // UpdateScoreboard should be called 8 times a second. 
```
So, rather make that a property. "UpdatePerSecond". Interval = 1000 / UpdatePerSecond

Lithrun commented on Assets/Scripts/UI/InGame/Scoreboard/Scoreboard.cs:
```diff
@@ -0,0 +1,195 @@
+
+            switch (sortLabel)
+            {
+                case "ID":
```
What about internationalization?

Lithrun commented on Assets/Scripts/UI/InGame/Scoreboard/Scoreboard.cs:
```diff
@@ -0,0 +1,195 @@
+    #endregion
+
+        // Formula for calculating score using svork balance
+        private static float CalculateScore(int k, int d, int m, int t)
```
Score should be calculated by a gamemode, as score depends on the gamemode. I.e. when someone plays capture gamemode, and they get 3 2k's, but haven't captured any objects, their score would be lower than someone who captured two objectives. For racing, the score would be the time, ect

Lithrun commented on Assets/Scripts/UI/InGame/Scoreboard/SortButton.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace Assets.Scripts.UI.InGame.Scoreboard
+{
+    public class SortButton : MonoBehaviour
```
PascalCasing & Access Modifiers

Lithrun commented on Assets/Scripts/UI/InGame/Scoreboard/SortButton.cs:
```diff
@@ -0,0 +1,55 @@
+            
+            GetComponent<Button>().onClick.AddListener(Sort);
+
+            scoreboard = GameObject.Find("Scoreboard").GetComponent<Scoreboard>();
```
UIService can get you a direct reference to the Scoreboard

Lithrun reviewed (commented):
Initial review

Commit:
AoTTG 3

new game

ps. just a duplicate of aottg2.unity for merge conflict fixes

Commit:
Merge branch 'development' into #537-revamp-ingame-hud

Commit:
Fixed conflict + proper file renaming

Lithrun committed:
Fix exceptions

Lithrun committed:
Fix .csproj files

Lithrun committed:
Fix namespace

Lithrun reviewed (changes requested):
As per commit: https://github.com/AoTTG-2/AoTTG-2/pull/607/commits/0e211d0901f42c10834afebb6b967b43961a46bf the CI is broken. Please resolve this. Once done, I can review it.

FAStre committed:
Change naming

FAStre committed:
Rename ChangeHUDHandler References

FAStre committed:
Rename ChangeHudHandler.cs properly

FAStre committed:
Add reference in inspector

Commit:
Renamed ChangeHUDHandler to ChangeHudHandler + fix reference errors

gisketch committed:
Delete ChangeHUDHandler.cs

gisketch committed:
Delete ChangeHUDHandler.cs.meta

gisketch committed:
Delete CustomizableHUDElement.cs

gisketch committed:
Delete CustomizableHUDElement.cs.meta

Commit:
Update AoTTG 2.unity

Commit:
Merge branch '#537-revamp-ingame-hud' of https://github.com/AoTTG-2/AoTTG-2 into #537-revamp-ingame-hud

Lithrun commented:
> 
> 
> As per commit: [0e211d0](https://github.com/AoTTG-2/AoTTG-2/commit/0e211d0901f42c10834afebb6b967b43961a46bf) the CI is broken. Please resolve this. Once done, I can review it.

Alright, thanks for the fixes @gisketch + @FAStre. Will fix a few code smells, and then get this one merged into development.

Lithrun committed:
Added minor improvements:
- AHSS & Blades HUD no longer collide with each other
- Code improvements & code smells
